base/loading.jl

```diff
@@ -3115,9 +3115,16 @@ end
 
         # now check if this file is fresh relative to its source files
         if !skip_timecheck
-            if !samefile(includes[1].filename, modpath) && !samefile(fixup_stdlib_path(includes[1].filename), modpath)
-                @debug "Rejecting cache file $cachefile because it is for file $(includes[1].filename) not file $modpath"
-                return true # cache file was compiled from a different path
+            if !samefile(includes[1].filename, modpath)
+                stdlib_path = fixup_stdlib_path(includes[1].filename)
+                # In certain cases the path rewritten by `fixup_stdlib_path` may
+                # point to an unreadable directory, make sure we can `stat` the
+                # file before comparing it with `modpath`.
+                isreadable = iszero(@ccall jl_fs_access(stdlib_path::Cstring, 0x04::Cint)::Cint)
+                if isreadable !samefile(stdlib_path, modpath)
```

Comment:
Doesn't this end up assuming fresh-ness, e.g., if the path from the include is not stat-able?

For the problem this is fixing, that's the correct thing to assume - but is it always?

Reply:
I was hoping @vchuravy (who wrote this condition) or @fatteneder (who should be somewhat familiar with the caching business) would have a good answer to this question 🙂

Reply:
@topolarity are you suggesting replacing the condition
```julia
isreadable && !samefile
```
with
```julia
!isreadable || !samefile
```
?

Reply:
> I was hoping @vchuravy (who wrote this condition) 

Uhm... I don't think I did.

I dislike when People apply "de Morgan's"

```
            if !samefile(includes[1].filename, modpath) && !samefile(fixup_stdlib_path(includes[1].filename), modpath)
```

Is 
```
            if !(samefile(includes[1].filename, modpath) || samefile(fixup_stdlib_path(includes[1].filename), modpath))
```

So if neither the filename nor the fixed up filename match, then we reject the cachefile. 


So, is the issue here that "samefile" requires the path to exist on the filesystem?


Our condition should be:
```
if !(samefile(includes[1].filename, modpath) ||
     (readable(fixup_stdlib_path(includes[1].filename)) && samefile(fixup_stdlib_path(includes[1].filename), modpath)))
```

Which would mean:
````
isreadable = iszero(@ccall jl_fs_access(stdlib_path::Cstring, 0x04::Cint)::Cint)
if !(isreadable && !samefile(stdlib_path, modpath))
    @debug "Rejecting cache file $cachefile because it is for file $(includes[1].filename) not file $modpath"
    return true # cache file was compiled from a different path
end
```

Since we want to reject if `isreadable == false` or if `isreadable` and it's not the samefile.

with an unhelpful de Morgan's theorem application:


`if !isreadable || samefile(stdlib_path, modpath))`

Reply:
> Uhm... I don't think I did.

It was introduced in your PR to add pkgimages:
https://github.com/JuliaLang/julia/pull/47184/files#diff-ec80afe9f8ff182765c73ed3fc35a103b561f873a52b1dfe1ce135e4e5f9a2cfR2610 🙂  but commits were squashed, so hard to know who exactly wrote that line.

> So, is the issue here that "samefile" requires the path to exist on the filesystem?

It needs the paths to be `stat`-able, rather then existing: it's fine with non-existing files:
```
julia> samefile("/does/not/exist", "/neither/does/this")
false
```
but if the parent directories are unreadable, then `samefile` (or `stat`) throws out `EACCES`

Reply:
> Which would mean:
> ```
> isreadable = iszero(@ccall jl_fs_access(stdlib_path::Cstring, 0x04::Cint)::Cint)
> if !(isreadable && !samefile(stdlib_path, modpath))
>     @debug "Rejecting cache file $cachefile because it is for file $(includes[1].filename) not file $modpath"
>     return true # cache file was compiled from a different path
> end
> ```

Did you mean
```
if !(isreadable && samefile(stdlib_path, modpath))
```
?

Reply:
We don't want to reject the cachefile (return true) if the files are the same.

Reply:
I'm probably being thick, but doesn't your snippet return `true` if files are the same?